File: leetCode/python/p206_ReverseLinkedList/lc206.py
#!/usr/bin/python3

# Definition for singly-linked list.
class ListNode:
	def __init__(self, val=0, next=None):
		self.val = val
		self.next = next

class Solution:
	'''
	Given the head of a singly linked list, reverse the list, and return the reversed list.
	'''
	def __init__(self):
		self.head = None

	def append_node(self, val):
		new_node = ListNode(val)
		# check if head already exists
		if self.head:
			current = self.head
			# go to the end of linked list
			while current.next:
				current = current.next
			current.next = new_node
		else:
			self.head = new_node

	def append_nodes_from_list(self, val_list: list):
		for i in range(len(val_list)):
			self.append_node(val_list[i])

	def print_linked_list(self):
		current = self.head
		
		# if list is empty print []	
		if current == None:
			print('[]', end = ' ')
	
		while current:
			print(current.val, end = ' ')
			current = current.next
		print()

	def reverseList(self) -> ListNode:
	
		# check if linked list is empty	
		if self.head == None:
			return self.head

		curr = self.head
		prev = None		
	
		while curr is not None:
			next = curr.next
			curr.next = prev
			prev = curr
			curr = next

		self.head = prev
	
		return self.head

# Test 1:

print('Test 1:')

list1 = Solution()
list1_array = [1, 2, 3, 4, 5]

list1.append_nodes_from_list(list1_array)

print('List 1 before reversing: ', end = ' ')
list1.print_linked_list()

list1.reverseList()

print('List 1 after reversing: ', end = ' ')
list1.print_linked_list()

# Test 2:

print('Test 2:')

list2 = Solution()
list2_array = [1, 2]

list2.append_nodes_from_list(list2_array)

print('List 2 before reversing: ', end = ' ')
list2.print_linked_list()

list2.reverseList()

print('List 2 after reversing: ', end = ' ')
list2.print_linked_list()

# Test 2:

print('Test 2:')

list3 = Solution()

print('List 3 before reversing: ', end = ' ')
list3.print_linked_list()

list3.reverseList()

print('List 3 after reversing: ', end = ' ')
list3.print_linked_list()

File: leetCode/c/p206_ReverseLinkedList/lc206.c
/* Given the head of a singly linked list, reverse the list, and return the reversed list. */

#include <stdio.h>
#include <stdlib.h>
#include "mylib.h"

int main(void) {

	/* Test 1: */
		
	struct ListNodeInt *list1, *list1Reversed;
	int list1Array[5] = {1, 2, 3, 4, 5};
	int list1ArraySize = sizeof(list1Array) / sizeof(int);

	listInitializeInt(&list1);
	listInsertArrayInt(list1, list1Array, list1ArraySize);		

	printf("List 1 before reversing: ");	
	listPrintInt(list1);

	list1Reversed = reverseListInt(list1);	

	printf("List 1 after reversing: ");	
	listPrintInt(list1Reversed);

	/* Test 2: */
	
	struct ListNodeInt *list2, *list2Reversed;
	int list2Array[2] = {1, 2};
	int list2ArraySize = sizeof(list2Array) / sizeof(int);
	
	listInitializeInt(&list2);
	listInsertArrayInt(list2, list2Array, list2ArraySize);		

	printf("List 2 before reversing: ");	
	listPrintInt(list2);

	list2Reversed = reverseListInt(list2);	
	
	printf("List 2 after reversing: ");	
	listPrintInt(list2Reversed);

	/* Test 3: */

	struct ListNodeInt *list3, *list3Reversed;

	listInitializeInt(&list3);

	printf("List 3 before reversing: ");	
	listPrintInt(list3);

	list3Reversed = reverseListInt(list3);

	printf("List 3 after reversing: ");	
	listPrintInt(list3Reversed);
	
	return EXIT_SUCCESS;
}
